File: backend-java/src/test/java/com/skewwhiffy/auraltester/notation/factory/ClefFactoryTest.java
package com.skewwhiffy.auraltester.notation.factory;

import com.skewwhiffy.auraltester.model.ClefType;
import com.skewwhiffy.auraltester.notation.model.clef.Clef;
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote;
import com.skewwhiffy.auraltester.test.util.TestData;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ClefFactoryTest {
    @Mock
    private NoteFactory noteFactory;
    @InjectMocks
    private ClefFactory clefFactory;

    private AbsoluteNote highLedgerNote;
    private AbsoluteNote lowLedgerNote;

    @BeforeEach
    void setUp() {
        highLedgerNote = TestData.random().absoluteNote();
        lowLedgerNote = TestData.random().absoluteNote();
    }

    @ParameterizedTest
    @MethodSource("clefTypeTestCases")
    void getFromClefTypeWorks(ClefType clefType, Function<ClefFactory, Clef> getExpected) {
        val expected = getExpected.apply(clefFactory);

        val actual = clefFactory.get(clefType);

        assertThat(actual).isEqualTo(expected);
    }

    @Test
    void initializedTrebleClefCorrectly() {
        when(noteFactory.getAbsoluteNote("a")).thenReturn(highLedgerNote);

        val actual = clefFactory.getTreble();

        assertThat(actual.abc()).isEqualTo("treble");
        assertThat(actual.lowLedgerNote()).isEqualTo(AbsoluteNote.getMiddleC());
        assertThat(actual.highLedgerNote()).isEqualTo(highLedgerNote);
    }

    @Test
    void initializesAltoClefCorrectly() {
        testGeneric("alto", "D,", "B", () -> clefFactory.getAlto());
    }

    @Test
    void initializesTenorClefCorrectly() {
        testGeneric("tenor", "B,,", "G", () -> clefFactory.getTenor());
    }

    @Test
    void initializesBassClefCorrectly() {
        when(noteFactory.getAbsoluteNote("E,,")).thenReturn(lowLedgerNote);

        val actual = clefFactory.getBass();

        assertThat(actual.abc()).isEqualTo("bass");
        assertThat(actual.lowLedgerNote()).isEqualTo(lowLedgerNote);
        assertThat(actual.highLedgerNote()).isEqualTo(AbsoluteNote.getMiddleC());
    }

    public static Stream<Arguments> clefTypeTestCases() {
        return Stream.of(
                Arguments.of(ClefType.TREBLE, (Function<ClefFactory, Clef>) ClefFactory::getTreble),
                Arguments.of(ClefType.ALTO, (Function<ClefFactory, Clef>) ClefFactory::getAlto),
                Arguments.of(ClefType.TENOR, (Function<ClefFactory, Clef>) ClefFactory::getTenor),
                Arguments.of(ClefType.BASS, (Function<ClefFactory, Clef>) ClefFactory::getBass)
        );
    }

    private void testGeneric(
            String abc,
            String expectedLowLedgerNote,
            String expectedHighLedgerNote,
            Supplier<Clef> getClef
    ) {
        when(noteFactory.getAbsoluteNote(expectedLowLedgerNote)).thenReturn(lowLedgerNote);
        when(noteFactory.getAbsoluteNote(expectedHighLedgerNote)).thenReturn(highLedgerNote);

        val actual = getClef.get();

        assertThat(actual.abc()).isEqualTo(abc);
        assertThat(actual.lowLedgerNote()).isEqualTo(lowLedgerNote);
        assertThat(actual.highLedgerNote()).isEqualTo(highLedgerNote);
    }
}

File: backend/src/test/kotlin/com/skewwhiffy/notation/factory/ClefFactoryTest.kt
package com.skewwhiffy.notation.factory

import com.skewwhiffy.notation.model.clef.Clef
import com.skewwhiffy.notation.model.note.AbsoluteNote
import com.skewwhiffy.test.util.TestData
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class ClefFactoryTest {
  @Mock
  private lateinit var noteFactory: NoteFactory
  @InjectMocks
  private lateinit var clefFactory: ClefFactory

  private lateinit var highLedgerNote: AbsoluteNote
  private lateinit var lowLedgerNote: AbsoluteNote

  @BeforeEach
  fun `set up`() {
    highLedgerNote = TestData.random.absoluteNote
    lowLedgerNote = TestData.random.absoluteNote
  }

  @Test
  fun`initialize treble clef correctly` () {
    `when`(noteFactory.getAbsoluteNote("a")).thenReturn(highLedgerNote)

    val actual = clefFactory.treble

    assertThat(actual.abc).isEqualTo("treble")
    assertThat(actual.lowLedgerNote).isEqualTo(AbsoluteNote.middleC)
    assertThat(actual.highLedgerNote).isEqualTo(highLedgerNote)
  }

  @Test
  fun`initialize alto clef correctly` () {
    testGeneric("alto", "D,", "B") {
      clefFactory.alto
    }
  }

  @Test
  fun`initialize tenor clef correctly` () {
    testGeneric("tenor", "B,,", "G") {
      clefFactory.tenor
    }
  }

  @Test
  fun`bass clef initializes correctly` () {
    `when`(noteFactory.getAbsoluteNote("E,,")).thenReturn(lowLedgerNote)

    val actual = clefFactory.bass

    assertThat(actual.abc).isEqualTo("bass")
    assertThat(actual.lowLedgerNote).isEqualTo(lowLedgerNote)
    assertThat(actual.highLedgerNote).isEqualTo(AbsoluteNote.middleC)
  }

  private fun testGeneric(
    abc: String,
    expectedLowLedgerNote: String,
    expectedHighLedgerNote: String,
    getClef: () -> Clef
  ) {
    `when`(noteFactory.getAbsoluteNote(expectedLowLedgerNote)).thenReturn(lowLedgerNote)
    `when`(noteFactory.getAbsoluteNote(expectedHighLedgerNote)).thenReturn(highLedgerNote)

    val actual = getClef()

    assertThat(actual.abc).isEqualTo(abc)
    assertThat(actual.lowLedgerNote).isEqualTo(lowLedgerNote)
    assertThat(actual.highLedgerNote).isEqualTo(highLedgerNote)
  }
}